rusk/src/lib/pow_verifier.rs

```diff
@@ -0,0 +1,76 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use blake2::{Blake2s256, Digest};
+
+pub const POW_DIFFICULTY: usize = 16;
+
+#[derive(Debug)]
+pub struct PoW;
+
+impl PoW {
+    /// Produces a nonce such that hash of a given collection of bytes
+    /// concatenated with the nonce has difficulty `difficulty`
+    /// (meaning, its last `difficulty` bits will be zero)
+    pub fn generate(bytes: impl AsRef<[u8]>, difficulty: usize) -> Vec<u8> {
+        let mut nonce: u64 = 0;
+        let mut hasher = Blake2s256::new();
+        loop {
+            hasher.update(bytes.as_ref());
+            let nonce_bytes = nonce.to_le_bytes().to_vec();
+            hasher.update(nonce_bytes.clone());
+            if Self::verify_difficulty(
+                &mut hasher.finalize_reset().iter().rev(),
+                difficulty,
+            ) {
+                return nonce_bytes;
+            }
+            nonce += 1;
+        }
+    }
+
+    /// Verifies the difficulty of a binary value according to the specified
+    /// criteria.
+    fn verify_difficulty<'a, I>(bytes: &mut I, difficulty: usize) -> bool
```

Comment:
Why are we using the trailing bits approach instead of the leading ones?
Using the latter would be as simple to verify as checking if the value of bytes is lower than the value of the string with `difficulty-1` initial bits to zero